File: harness/platform_core_contract/verification.py
"""Pure Platform Core VerificationRequest/Receipt v1 validation."""

from __future__ import annotations

from typing import Any

from .artifact import (_validate_artifact_references,
                       _validate_artifact_relations,
                       _validate_artifact_values)
from .codec import (ContractError, canonical_json, decode_canonical, exact_object,
                    lower_token, observation_digest, validate_hash, validate_i64,
                    validate_schema_name, validate_document_shape, validate_text)
from .constants import (CANONICALIZATION, MAX_EVIDENCE_REFS,
                        DOCUMENT_INVALID, MAX_RECEIPT_BYTES, MAX_UNIX_MILLISECONDS,
                        MAX_VERIFICATION_CHECKS, RECEIPT_VERSION,
                        REFERENCE_MISMATCH, RELATION_MISMATCH, STATE_INVALID,
                        VERIFICATION_CHECK_REQUEST_FIELDS,
                        VERIFICATION_RECEIPT_TYPE_PROFILE,
                        VERIFICATION_RECEIPT_FIELDS,
                        VERIFICATION_REQUEST_DIGEST_DOMAIN,
                        VERIFICATION_REQUEST_FIELDS,
                        VERIFICATION_REQUEST_TYPE_PROFILE,
                        VERIFICATION_RESULT_FIELDS)
from .execution_receipt import validate_reason_codes
from .identity import validate_typed_id
from .references import (validate_actor_ref, validate_record_ref,
                         validate_scope_references, validate_scope_values)

_STATUSES = {"pass", "fail", "inconclusive", "not_executed"}
_SEVERITY = {"pass": 0, "not_executed": 1, "inconclusive": 2, "fail": 3}


def validate_verification_request(value: Any) -> dict[str, Any]:
    """Return one valid immutable-input verification request."""
    validate_document_shape(value, VERIFICATION_REQUEST_TYPE_PROFILE,
                            "VerificationRequest")
    node = exact_object(value, VERIFICATION_REQUEST_FIELDS,
                        "VerificationRequest")
    _validate_request_values(node)
    canonical_json(node, MAX_RECEIPT_BYTES)
    _validate_input_references(node["scope_ref"], node["input_artifact_ref"])
    _validate_request_relations(node)
    return node


def _validate_request_values(node: dict[str, Any]) -> None:
    if (node["canonicalization"] != CANONICALIZATION or
            validate_i64(node["verification_request_version"],
                         "verification_request_version", 1) != RECEIPT_VERSION):
        raise ContractError("verification request version or canonicalization is unsupported")
    validate_typed_id(node["verification_id"], "ver", "verification_id")
    validate_actor_ref(node["requested_by"])
    _unix_ms(node["requested_at_unix_ms"], "requested_at_unix_ms")
    validate_scope_values(node["scope_ref"])
    _validate_artifact_values(node["input_artifact_ref"])
    _validate_check_requests(node["checks"])


def _validate_request_relations(node: dict[str, Any]) -> None:
    _validate_artifact_relations(node["input_artifact_ref"])
    if node["input_artifact_ref"]["created_at_unix_ms"] > node["requested_at_unix_ms"]:
        raise ContractError("verification input cannot postdate its request",
                            RELATION_MISMATCH)


def validate_verification_receipt(value: Any) -> dict[str, Any]:
    """Return one valid supplied verification observation declaration."""
    validate_document_shape(value, VERIFICATION_RECEIPT_TYPE_PROFILE,
                            "VerificationReceipt")
    node = exact_object(value, VERIFICATION_RECEIPT_FIELDS,
                        "VerificationReceipt")
    _validate_receipt_values(node)
    canonical_json(node, MAX_RECEIPT_BYTES)
    _validate_input_references(node["scope_ref"], node["input_artifact_ref"])
    _validate_receipt_state(node)
    _validate_receipt_relations(node)
    return node


def _validate_receipt_values(node: dict[str, Any]) -> None:
    if (node["canonicalization"] != CANONICALIZATION or
            validate_i64(node["verification_receipt_version"],
                         "verification_receipt_version", 1) != RECEIPT_VERSION):
        raise ContractError("verification receipt version or canonicalization is unsupported")
    validate_typed_id(node["receipt_id"], "rcp", "receipt_id")
    validate_typed_id(node["verification_id"], "ver", "verification_id")
    validate_hash(node["request_sha256"], "request_sha256")
    _validate_producer(node["produced_by"])
    validate_scope_values(node["scope_ref"])
    _validate_artifact_values(node["input_artifact_ref"])
    _unix_ms(node["started_at_unix_ms"], "started_at_unix_ms")
    _unix_ms(node["ended_at_unix_ms"], "ended_at_unix_ms")
    _validate_check_result_values(node["results"])


def validate_verification_exchange(request: Any, receipt: Any) -> None:
    """Compare two already supplied records without resolving evidence."""
    request_node = _detached_exchange_value(
        request, VERIFICATION_REQUEST_TYPE_PROFILE, "VerificationRequest",
        VERIFICATION_REQUEST_FIELDS)
    receipt_node = _detached_exchange_value(
        receipt, VERIFICATION_RECEIPT_TYPE_PROFILE, "VerificationReceipt",
        VERIFICATION_RECEIPT_FIELDS)
    _validate_request_values(request_node)
    _validate_receipt_values(receipt_node)
    _validate_input_references(request_node["scope_ref"],
                               request_node["input_artifact_ref"])
    _validate_input_references(receipt_node["scope_ref"],
                               receipt_node["input_artifact_ref"])
    _validate_receipt_state(receipt_node)
    _validate_request_relations(request_node)
    _validate_receipt_relations(receipt_node)
    _validate_exchange_relations(request_node, receipt_node)


def _validate_exchange_relations(request_node: dict[str, Any],
                                 receipt_node: dict[str, Any]) -> None:
    request_digest = observation_digest(
        VERIFICATION_REQUEST_DIGEST_DOMAIN,
        canonical_json(request_node, MAX_RECEIPT_BYTES))
    if (request_node["verification_id"] != receipt_node["verification_id"] or
            receipt_node["request_sha256"] != request_digest):
        raise ContractError("verification receipt does not bind the exact request",
                            RELATION_MISMATCH)
    if (request_node["scope_ref"] != receipt_node["scope_ref"] or
            request_node["input_artifact_ref"] != receipt_node["input_artifact_ref"]):
        raise ContractError("verification request and receipt scope or input differs",
                            RELATION_MISMATCH)
    if receipt_node["started_at_unix_ms"] < request_node["requested_at_unix_ms"]:
        raise ContractError("verification cannot start before its request",
                            RELATION_MISMATCH)
    _compare_checks(request_node["checks"], receipt_node["results"])


def _detached_exchange_value(value: Any, profile: Any, label: str,
                             fields: set[str]) -> dict[str, Any]:
    if type(value) is not dict:
        raise ContractError(f"{label} must be an object", DOCUMENT_INVALID)
    canonical = canonical_json(value, MAX_RECEIPT_BYTES)
    snapshot = decode_canonical(canonical, MAX_RECEIPT_BYTES)
    validate_document_shape(snapshot, profile, label)
    return exact_object(snapshot, fields, label)


def _validate_input_references(scope_value: Any, input_value: Any) -> None:
    scope = scope_value
    validate_scope_references(scope)
    _validate_artifact_references(input_value)
    if scope["project_snapshot_id"] is None or scope["attempt_id"] is None:
        raise ContractError("verification scope requires project snapshot and attempt",
                            REFERENCE_MISMATCH)
    artifact = input_value
    if (artifact["source_snapshot_ref"]["entity_id"] != scope["project_snapshot_id"] or
            artifact["producer_attempt_id"] != scope["attempt_id"]):
        raise ContractError("verification input must match scoped snapshot and attempt",
                            REFERENCE_MISMATCH)


def _validate_check_requests(values: Any) -> None:
    if type(values) is not list:
        raise ContractError("checks must be an array", DOCUMENT_INVALID)
    if not 1 <= len(values) <= MAX_VERIFICATION_CHECKS:
        raise ContractError(
            f"checks cardinality must be 1..{MAX_VERIFICATION_CHECKS}")
    previous = ""
    for value in values:
        node = exact_object(value, VERIFICATION_CHECK_REQUEST_FIELDS,
                            "VerificationCheckRequest")
        _check_id(node["check_id"])
        if node["check_id"] <= previous:
            raise ContractError("checks must be strictly sorted and unique by check_id")
        validate_schema_name(node["check_name"], "check_name")
        if type(node["declared_required"]) is not bool:
            raise ContractError("declared_required must be boolean", DOCUMENT_INVALID)
        previous = node["check_id"]


def _validate_producer(value: Any) -> None:
    actor = validate_actor_ref(value)
    if actor["actor_type"] != "harness":
        raise ContractError(
            "verification receipt producer must declare harness actor_type")


def _validate_receipt_relations(node: dict[str, Any]) -> None:
    _validate_artifact_relations(node["input_artifact_ref"])
    started = node["started_at_unix_ms"]
    ended = node["ended_at_unix_ms"]
    if ended < started or node["input_artifact_ref"]["created_at_unix_ms"] > started:
        raise ContractError("verification interval or input timing is invalid",
                            RELATION_MISMATCH)
    for result in node["results"]:
        _validate_check_result_relations(result)
    derived = _derive_verification_status(node["results"])
    if node["overall_status"] != derived:
        raise ContractError(f"overall_status must be derived as {derived}",
                            RELATION_MISMATCH)


def _validate_receipt_state(node: dict[str, Any]) -> None:
    _validate_status(node["overall_status"], "overall_status")
    for result in node["results"]:
        _validate_status(result["status"], "verification status")


def _validate_check_result_values(values: Any) -> None:
    if type(values) is not list:
        raise ContractError("results must be an array", DOCUMENT_INVALID)
    if not 1 <= len(values) <= MAX_VERIFICATION_CHECKS:
        raise ContractError(
            f"results cardinality must be 1..{MAX_VERIFICATION_CHECKS}")
    previous = ""
    for value in values:
        node = _validate_check_result_value(value)
        if node["check_id"] <= previous:
            raise ContractError("results must be strictly sorted and unique by check_id")
        previous = node["check_id"]


def _validate_check_result_value(value: Any) -> dict[str, Any]:
    node = exact_object(value, VERIFICATION_RESULT_FIELDS,
                        "VerificationCheckResult")
    _check_id(node["check_id"])
    applicability = node["applicability"]
    if applicability not in {"applicable", "not_applicable"}:
        raise ContractError(f"applicability {applicability!r} is unsupported")
    if node["applicability_reason"] is not None:
        validate_text(node["applicability_reason"], "applicability_reason", 512)
    validate_reason_codes(node["reason_codes"], "result.reason_codes")
    _validate_evidence_refs(node["evidence_refs"])
    return node


def _validate_check_result_relations(node: dict[str, Any]) -> None:
    _validate_applicability(node)
    if (node["status"] == "pass") != (len(node["reason_codes"]) == 0):
        raise ContractError("pass requires no reasons; other statuses require reasons",
                            RELATION_MISMATCH)


def _derive_verification_status(values: list[dict[str, Any]]) -> str:
    overall, applicable = "pass", False
    for value in values:
        if value["applicability"] == "applicable":
            applicable = True
            if _SEVERITY[value["status"]] > _SEVERITY[overall]:
                overall = value["status"]
    return overall if applicable else "not_executed"


def _validate_status(value: Any, label: str) -> None:
    if type(value) is not str:
        raise ContractError(f"{label} must be text", DOCUMENT_INVALID)
    if value not in _STATUSES:
        raise ContractError(f"{label} {value!r} is unsupported", STATE_INVALID)


def _validate_applicability(node: dict[str, Any]) -> None:
    applicability, reason = node["applicability"], node["applicability_reason"]
    if type(applicability) is not str:
        raise ContractError("applicability must be text", DOCUMENT_INVALID)
    if applicability == "applicable":
        if reason is not None:
            raise ContractError("applicable result requires null applicability_reason",
                                RELATION_MISMATCH)
    elif applicability == "not_applicable":
        if node["status"] != "not_executed" or reason is None:
            raise ContractError("not_applicable requires not_executed and a reason",
                                RELATION_MISMATCH)
        validate_text(reason, "applicability_reason", 512)
    else:
        raise ContractError(f"applicability {applicability!r} is unsupported")


def _validate_evidence_refs(values: Any) -> None:
    if type(values) is not list:
        raise ContractError("evidence_refs must be an array", DOCUMENT_INVALID)
    if len(values) > MAX_EVIDENCE_REFS:
        raise ContractError(
            f"evidence_refs must have at most {MAX_EVIDENCE_REFS} items")
    previous = ""
    for value in values:
        reference = validate_record_ref(value, "evidence_ref")
        if reference["record_id"] <= previous:
            raise ContractError("evidence_refs must be strictly sorted and unique")
        previous = reference["record_id"]


def _compare_checks(requests: list[dict[str, Any]],
                    results: list[dict[str, Any]]) -> None:
    if (len(requests) != len(results) or
            any(request["check_id"] != result["check_id"]
                for request, result in zip(requests, results))):
        raise ContractError(
            "verification result set must exactly cover requested checks",
            RELATION_MISMATCH)


def _check_id(value: Any) -> None:
    if type(value) is not str:
        raise ContractError("check_id must be text", DOCUMENT_INVALID)
    if len(value) > 64 or not lower_token(value):
        raise ContractError("check_id must be a bounded lowercase token")


def _unix_ms(value: Any, label: str) -> int:
    return validate_i64(value, label, 1, MAX_UNIX_MILLISECONDS)

File: forge-runtime/crates/domain/src/platform_core_contract/verification.rs
use super::{
    ActorRef, ActorType, ArtifactRef, CANONICALIZATION, MAX_EVIDENCE_REFS, MAX_RECEIPT_BYTES,
    MAX_UNIX_MILLISECONDS, MAX_VERIFICATION_CHECKS, PlatformCoreContractError, RECEIPT_VERSION,
    RejectionCode, VerificationApplicability, VerificationCheckRequest, VerificationCheckResult,
    VerificationReceipt, VerificationRequest, VerificationStatus, artifact, codec,
    execution_receipt, identity, references, reject, wire,
};

/// Validates one supplied immutable-input verification request.
///
/// # Errors
/// Returns a stable coded error for invalid declarations, references, or bounds.
pub fn validate_verification_request(
    value: &VerificationRequest,
) -> Result<(), PlatformCoreContractError> {
    validate_request_values(value)?;
    wire::canonical_typed(value, MAX_RECEIPT_BYTES)?;
    validate_input_references(&value.scope_ref, &value.input_artifact_ref)?;
    validate_request_relations(value)
}

fn validate_request_values(value: &VerificationRequest) -> Result<(), PlatformCoreContractError> {
    if value.canonicalization != CANONICALIZATION
        || value.verification_request_version != RECEIPT_VERSION
    {
        return Err(reject(
            RejectionCode::ValueInvalid,
            "verification request version or canonicalization is unsupported",
        ));
    }
    identity::validate_typed_id(&value.verification_id, "ver", "verification_id")?;
    references::validate_actor_ref(&value.requested_by)?;
    validate_unix_ms(value.requested_at_unix_ms, "requested_at_unix_ms")?;
    references::validate_scope_values(&value.scope_ref)?;
    artifact::validate_artifact_values(&value.input_artifact_ref)?;
    validate_check_requests(&value.checks)
}

fn validate_request_relations(
    value: &VerificationRequest,
) -> Result<(), PlatformCoreContractError> {
    artifact::validate_artifact_relations(&value.input_artifact_ref)?;
    if value.input_artifact_ref.created_at_unix_ms > value.requested_at_unix_ms {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "verification input cannot postdate its request",
        ));
    }
    Ok(())
}

/// Validates supplied verification observations without declaring product completion.
///
/// # Errors
/// Returns a stable coded error for invalid declarations, references, state, or relations.
pub fn validate_verification_receipt(
    value: &VerificationReceipt,
) -> Result<(), PlatformCoreContractError> {
    validate_receipt_values(value)?;
    wire::canonical_typed(value, MAX_RECEIPT_BYTES)?;
    validate_input_references(&value.scope_ref, &value.input_artifact_ref)?;
    validate_receipt_state(value)?;
    validate_receipt_relations(value)
}

fn validate_receipt_values(value: &VerificationReceipt) -> Result<(), PlatformCoreContractError> {
    validate_receipt_header(value)?;
    validate_producer(&value.produced_by)?;
    references::validate_scope_values(&value.scope_ref)?;
    artifact::validate_artifact_values(&value.input_artifact_ref)?;
    validate_unix_ms(value.started_at_unix_ms, "started_at_unix_ms")?;
    validate_unix_ms(value.ended_at_unix_ms, "ended_at_unix_ms")?;
    validate_check_result_values(&value.results)
}

/// Compares one exact request and receipt without resolving referenced evidence.
///
/// # Errors
/// Returns `pc_relation_mismatch` when the two valid records do not bind exactly.
pub fn validate_verification_exchange(
    request: &VerificationRequest,
    receipt: &VerificationReceipt,
) -> Result<(), PlatformCoreContractError> {
    wire::canonical_typed(request, MAX_RECEIPT_BYTES)?;
    wire::canonical_typed(receipt, MAX_RECEIPT_BYTES)?;
    validate_request_values(request)?;
    validate_receipt_values(receipt)?;
    validate_input_references(&request.scope_ref, &request.input_artifact_ref)?;
    validate_input_references(&receipt.scope_ref, &receipt.input_artifact_ref)?;
    validate_receipt_state(receipt)?;
    validate_request_relations(request)?;
    validate_receipt_relations(receipt)?;
    validate_exchange_relations(request, receipt)
}

fn validate_exchange_relations(
    request: &VerificationRequest,
    receipt: &VerificationReceipt,
) -> Result<(), PlatformCoreContractError> {
    let digest = codec::verification_request_sha256(request)?;
    if request.verification_id != receipt.verification_id || receipt.request_sha256 != digest {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "verification receipt does not bind the exact request",
        ));
    }
    if request.scope_ref != receipt.scope_ref
        || request.input_artifact_ref != receipt.input_artifact_ref
    {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "verification request and receipt scope or input differs",
        ));
    }
    if receipt.started_at_unix_ms < request.requested_at_unix_ms {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "verification cannot start before its request",
        ));
    }
    compare_checks(&request.checks, &receipt.results)
}

fn validate_receipt_header(value: &VerificationReceipt) -> Result<(), PlatformCoreContractError> {
    if value.canonicalization != CANONICALIZATION
        || value.verification_receipt_version != RECEIPT_VERSION
    {
        return Err(reject(
            RejectionCode::ValueInvalid,
            "verification receipt version or canonicalization is unsupported",
        ));
    }
    identity::validate_typed_id(&value.receipt_id, "rcp", "receipt_id")?;
    identity::validate_typed_id(&value.verification_id, "ver", "verification_id")?;
    wire::validate_hash(&value.request_sha256, "request_sha256")
}

fn validate_producer(value: &ActorRef) -> Result<(), PlatformCoreContractError> {
    references::validate_actor_ref(value)?;
    if value.actor_type == ActorType::Harness {
        Ok(())
    } else {
        Err(reject(
            RejectionCode::ValueInvalid,
            "verification receipt producer must declare harness actor_type",
        ))
    }
}

fn validate_input_references(
    scope: &super::ScopeRef,
    input: &ArtifactRef,
) -> Result<(), PlatformCoreContractError> {
    references::validate_scope_references(scope)?;
    artifact::validate_artifact_references(input)?;
    let (Some(snapshot), Some(attempt)) = (
        scope.project_snapshot_id.as_deref(),
        scope.attempt_id.as_deref(),
    ) else {
        return Err(reject(
            RejectionCode::ReferenceMismatch,
            "verification scope requires project snapshot and attempt",
        ));
    };
    if input.source_snapshot_ref.entity_id != snapshot || input.producer_attempt_id != attempt {
        return Err(reject(
            RejectionCode::ReferenceMismatch,
            "verification input must match scoped snapshot and attempt",
        ));
    }
    Ok(())
}

fn validate_check_requests(
    values: &[VerificationCheckRequest],
) -> Result<(), PlatformCoreContractError> {
    if values.is_empty() || values.len() > MAX_VERIFICATION_CHECKS {
        return Err(reject(
            RejectionCode::ValueInvalid,
            format!("checks cardinality must be 1..{MAX_VERIFICATION_CHECKS}"),
        ));
    }
    let mut previous = "";
    for value in values {
        validate_check_id(&value.check_id)?;
        if value.check_id.as_str() <= previous {
            return Err(reject(
                RejectionCode::ValueInvalid,
                "checks must be strictly sorted and unique by check_id",
            ));
        }
        wire::validate_schema_name(&value.check_name, "check_name")?;
        previous = &value.check_id;
    }
    Ok(())
}

fn validate_receipt_relations(
    value: &VerificationReceipt,
) -> Result<(), PlatformCoreContractError> {
    artifact::validate_artifact_relations(&value.input_artifact_ref)?;
    if value.ended_at_unix_ms < value.started_at_unix_ms
        || value.input_artifact_ref.created_at_unix_ms > value.started_at_unix_ms
    {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "verification interval or input timing is invalid",
        ));
    }
    for result in &value.results {
        validate_check_result_relations(result)?;
    }
    let derived = derive_verification_status(&value.results);
    if value.overall_status != derived {
        return Err(reject(
            RejectionCode::RelationMismatch,
            format!("overall_status must be derived as {derived:?}"),
        ));
    }
    Ok(())
}

fn validate_receipt_state(value: &VerificationReceipt) -> Result<(), PlatformCoreContractError> {
    validate_status(&value.overall_status, "overall_status")?;
    for result in &value.results {
        validate_status(&result.status, "verification status")?;
    }
    Ok(())
}

fn validate_check_result_values(
    values: &[VerificationCheckResult],
) -> Result<(), PlatformCoreContractError> {
    if values.is_empty() || values.len() > MAX_VERIFICATION_CHECKS {
        return Err(reject(
            RejectionCode::ValueInvalid,
            format!("results cardinality must be 1..{MAX_VERIFICATION_CHECKS}"),
        ));
    }
    let mut previous = "";
    for value in values {
        validate_check_result_value(value)?;
        if value.check_id.as_str() <= previous {
            return Err(reject(
                RejectionCode::ValueInvalid,
                "results must be strictly sorted and unique by check_id",
            ));
        }
        previous = &value.check_id;
    }
    Ok(())
}

fn validate_check_result_value(
    value: &VerificationCheckResult,
) -> Result<(), PlatformCoreContractError> {
    validate_check_id(&value.check_id)?;
    validate_applicability_value(&value.applicability)?;
    if let Some(reason) = value.applicability_reason.as_deref() {
        wire::validate_text(reason, "applicability_reason", 512, true)?;
    }
    execution_receipt::validate_reason_codes(&value.reason_codes, "result.reason_codes")?;
    validate_evidence_refs(&value.evidence_refs)
}

fn validate_check_result_relations(
    value: &VerificationCheckResult,
) -> Result<(), PlatformCoreContractError> {
    validate_applicability(value)?;
    if (value.status == VerificationStatus::Pass) != value.reason_codes.is_empty() {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "pass requires no reasons; other statuses require reasons",
        ));
    }
    Ok(())
}

fn derive_verification_status(values: &[VerificationCheckResult]) -> VerificationStatus {
    let mut overall = VerificationStatus::Pass;
    let mut applicable = false;
    for value in values {
        if value.applicability == VerificationApplicability::Applicable {
            applicable = true;
            overall = more_severe(overall, &value.status);
        }
    }
    if applicable {
        overall
    } else {
        VerificationStatus::NotExecuted
    }
}

fn validate_applicability(
    value: &VerificationCheckResult,
) -> Result<(), PlatformCoreContractError> {
    match &value.applicability {
        VerificationApplicability::Applicable if value.applicability_reason.is_none() => Ok(()),
        VerificationApplicability::NotApplicable
            if value.status == VerificationStatus::NotExecuted
                && value.applicability_reason.is_some() =>
        {
            wire::validate_text(
                value.applicability_reason.as_deref().unwrap_or_default(),
                "applicability_reason",
                512,
                true,
            )
        }
        VerificationApplicability::Applicable => Err(reject(
            RejectionCode::RelationMismatch,
            "applicable result requires null applicability_reason",
        )),
        VerificationApplicability::NotApplicable => Err(reject(
            RejectionCode::RelationMismatch,
            "not_applicable requires not_executed and a reason",
        )),
        VerificationApplicability::Unknown(_) => Err(reject(
            RejectionCode::ValueInvalid,
            "applicability is unsupported",
        )),
    }
}

fn validate_evidence_refs(values: &[super::RecordRef]) -> Result<(), PlatformCoreContractError> {
    if values.len() > MAX_EVIDENCE_REFS {
        return Err(reject(
            RejectionCode::ValueInvalid,
            format!("evidence_refs exceeds {MAX_EVIDENCE_REFS} items"),
        ));
    }
    let mut previous = "";
    for value in values {
        references::validate_record_ref(value, "evidence_ref")?;
        if value.record_id.as_str() <= previous {
            return Err(reject(
                RejectionCode::ValueInvalid,
                "evidence_refs must be strictly sorted and unique",
            ));
        }
        previous = &value.record_id;
    }
    Ok(())
}

fn validate_check_id(value: &str) -> Result<(), PlatformCoreContractError> {
    if wire::lower_token(value) && value.len() <= 64 {
        Ok(())
    } else {
        Err(reject(
            RejectionCode::ValueInvalid,
            "check_id must be a bounded lowercase token",
        ))
    }
}

fn more_severe(left: VerificationStatus, right: &VerificationStatus) -> VerificationStatus {
    if status_severity(right) > status_severity(&left) {
        right.clone()
    } else {
        left
    }
}

fn status_severity(value: &VerificationStatus) -> u8 {
    match value {
        VerificationStatus::Pass | VerificationStatus::Unknown(_) => 0,
        VerificationStatus::NotExecuted => 1,
        VerificationStatus::Inconclusive => 2,
        VerificationStatus::Fail => 3,
    }
}

fn validate_status(
    value: &VerificationStatus,
    label: &str,
) -> Result<(), PlatformCoreContractError> {
    if matches!(value, VerificationStatus::Unknown(_)) {
        Err(reject(
            RejectionCode::StateInvalid,
            format!("{label} is unsupported"),
        ))
    } else {
        Ok(())
    }
}

fn validate_applicability_value(
    value: &VerificationApplicability,
) -> Result<(), PlatformCoreContractError> {
    if matches!(value, VerificationApplicability::Unknown(_)) {
        Err(reject(
            RejectionCode::ValueInvalid,
            "applicability is unsupported",
        ))
    } else {
        Ok(())
    }
}

fn compare_checks(
    requests: &[VerificationCheckRequest],
    results: &[VerificationCheckResult],
) -> Result<(), PlatformCoreContractError> {
    if requests.len() == results.len()
        && requests
            .iter()
            .zip(results)
            .all(|(request, result)| request.check_id == result.check_id)
    {
        Ok(())
    } else {
        Err(reject(
            RejectionCode::RelationMismatch,
            "verification result set must exactly cover requested checks",
        ))
    }
}

fn validate_unix_ms(value: i64, label: &str) -> Result<(), PlatformCoreContractError> {
    if (1..=MAX_UNIX_MILLISECONDS).contains(&value) {
        Ok(())
    } else {
        Err(reject(
            RejectionCode::ValueInvalid,
            format!("{label} is outside the supported range"),
        ))
    }
}
